torch/csrc/lazy/ts_backend/ts_node.cpp

```diff
@@ -17,29 +17,29 @@ namespace torch {
 namespace lazy {
 
 
-hash_t OperandHashes(const OpList& operands, const hash_t& seed, bool bakeInSizes) {
+hash_t OperandHashes(const OpList& operands,
+                     const c10::ArrayRef<Shape>& shapes,
+                     const hash_t& seed, bool bakeInSizes) {
   hash_t hash = seed;
   for (auto& operand : operands) {
     if (!operand) {
       hash = HashCombine(hash, static_cast<uint64_t>(kNullOpt));
       continue;
     }
-    auto operand_hash = operand.hash();
+    auto operand_hash = bakeInSizes ? operand.shapeHash() : operand.hash();
     hash = HashCombine(hash, operand_hash);
   }
+  for (auto& shape : shapes) {
+    hash = HashCombine(hash, shape.hash(bakeInSizes));
+  }
   return hash;
 }
 
-hash_t GetOpHash(OpKind op, const Shape& shape, hash_t hash_seed, bool bakeInSizes) {
-  hash_t h = HashCombine(op.hash(), shape.hash(bakeInSizes));
-  return HashCombine(h, hash_seed);
-}
-
 TsNode::TsNode(OpKind op, OpList operands, std::vector<Shape>&& shapes, size_t num_outputs, hash_t hash_seed)
     : Node(op, operands, std::move(shapes), num_outputs) {
   hash_seed = HashCombine(op.hash(), hash_seed);
-  shape_hash_ = OperandHashes(operands, hash_seed, true);
-  dag_hash_ = (enableDynamicShape() ? OperandHashes(operands, hash_seed, false) : shape_hash_);
+  shape_hash_ = OperandHashes(operands, this->shapes(), hash_seed, true);
+  dag_hash_ = (enableDynamicShape() ? OperandHashes(operands, this->shapes(), hash_seed, false) : shape_hash_);
```

Comment:
The 2 hash handling looks correct to me.

I saw you passed in this->shapes() as well. Looks like we miss the current ops's shape information in the hash before?

Reply:
It was included in one of the hash functions, but not the other. So, I just combined to two into a single hash function which makes use of all of the passed in shapes.

Reply:
I think the reason this->shapes() is not passed to OperandHashes is, given OpKind and OpList, the shape of the output(the current) tensor should be deterministic. Thus we don't need to hash it's shape explicitly.

If there are no other reasons, probably we should keep the old behavior. cc @wconstab

Reply:
> Thus we don't need to hash it's shape explicitly.

Is your concern that we'll unnecessarily get more cache misses with this? Because I don't think this would affect that. Only ops that don't use the shape cache will have their shape hashes as part of the dag/node hash.

I can kind of see the concern with changing this here as it does affect all of the currently codegen'd native IR nodes. However, once the work is done to make those use the shape cache, I think this will not be a problem.

> given OpKind and OpList, the shape of the output(the current) tensor should be deterministic.

This wasn't the case with some ops that were using this hash function
https://github.com/pytorch/pytorch/blob/6066e5929f12cdef52fe5d85af8001ad3583ec42/torch/csrc/lazy/ts_backend/ts_node.cpp#L33-L36
included the shape hash as well.

Reply:
okay, maybe let's take a step back, why we need change
```
shape_hash_ = OperandHashes(operands,  hash_seed, true);
```
to
```
shape_hash_ = OperandHashes(operands, this->shapes(), hash_seed, true);
```
?

Does the previous implementation cause any problem?  I think your change should work as well. I just want to understand the motivation here.

Reply:
> this->shapes() is not passed to OperandHashes is, given OpKind and OpList, the shape of the output(the current) tensor should be deterministic

Yes, exactly.

This is only untrue for dynamic ops we do not yet support such as 'nonzero'.  For all the currently supported ops, the output shape is a deterministic function of the inputs to the op, for all ops except for the 'leaf' ops.  That's why shape was only hashed on leaf ops and then implicitly propagated thereafter.

I think IIUC antonio's change would add redundant information to the hash.  This means I don't expect any behavior change, but there would be a computational cost associated with the extra hashing.  Anyone think there is actually a behavior change (which I am missing?)

Reply:
I agree with the analysis above.

Reply:
As far as I've seen it doesn't affect the behaviour. The tests appear to confirm that. However, there is that extra hashing cost. Though, if we make the `LazyNativeFunctions` use the shape cache instead of computing the shapes every time, this extra cost disappears.

So, if we're okay with adding the extra cost now with the plan of removing this cost down the line, I'd prefer to keep it as is. However, if there is strong preference to not do this, I can bring back the old hash functions.

The reason I'd prefer to keep it the way it is that I made the non-native IR nodes return whatever their shape inference function returns (i.e. a vector of shapes). So, it will go to the first constructor instead of the leaf constructor. For those ops we need the shapes made as part of the hash. I also think it keeps things cleaner having a single hash function.

@wconstab thoughts?

Reply:
I think this is OK. If we find we need to optimize it, we have several paths to doing so.